### ISSUE TITLE
DataStore SQLiteException: Ambiguous Column Name

### ISSUE THREAD

**Issue by infernal008** [aws-amplify/amplify-android#1268]
I have a schema like below:

_I trimmed down most unecessary fields at it seems too verbose_

```
type CategoryObject @model {
  id: ID!
  transactions: [TransactionObject] @connection(keyName: "byCategory", fields: ["id"])
}

type WalletObject @model {
  id: ID!
  transactionsFrom: [TransactionObject] @connection(keyName: "byWalletFrom", fields: ["id"])
  transactionsTo: [TransactionObject] @connection(keyName: "byWalletTo", fields: ["id"])
}

type TransactionObject @model 
  @key(name: "byCategory", fields: ["categoryId"])
  @key(name: "byWalletFrom", fields: ["walletFromId"])
  @key(name: "byWalletTo", fields: ["walletToId"]){
  id: ID!
  categoryId: ID!
  category: CategoryObject @connection(fields: ["categoryId"])
  walletFromId: ID!
  walletFrom: WalletObject @connection(fields: ["walletFromId"])
  walletToId: ID!
  walletTo: WalletObject @connection(fields: ["walletToId"])
}
```
Basically:

- A TransactionObject has one CategoryObject
- A CategoryObject can be used by different TransactionObjects
- A TransactionObject has one WalletObject for walletFrom (in business logic: what wallet was used for the transaction)
- A TransactionObject can have a WalletObject for walletTo (in business logic: in what wallet will the transaction go, optional)
- A WalletObject can be used as either WalletFrom or WalletTo by TransactionObject (based on business logic)

I was able to provision this schema to my AWS instances via the Amplify CLI and was able to check if the models and relationships are working through the AppSync Console. It seems fine, I was able to create objects for all three models and relationships seem fine on query.

However, upon running the command await Amplify.DataStore.query(TransactionObject.classType) locally, I am blocked by a SQLiteException for Ambiguous Column Name:

```
E/amplify:flutter:datastore( 6792): Caused by: android.database.sqlite.SQLiteException: ambiguous column name: WalletObject.id (code 1 SQLITE_ERROR): , while compiling: SELECT `TransactionObject`.`id` AS `TransactionObject_id`, `TransactionObject`.`categoryId` AS `TransactionObject_categoryId`, `TransactionObject`.`walletFromId` AS `TransactionObject_walletFromId`, `TransactionObject`.`walletToId` AS `TransactionObject_walletToId`, `CategoryObject`.`id` AS `CategoryObject_id`, `WalletObject`.`id` AS `WalletObject_id`, `WalletObject`.`id` AS `WalletObject_id` FROM `TransactionObject` LEFT JOIN `CategoryObject` ON `TransactionObject`.`categoryId`=`CategoryObject`.`id` LEFT JOIN `WalletObject` ON `TransactionObject`.`walletFromId`=`WalletObject`.`id` LEFT JOIN `WalletObject` ON `TransactionObject`.`walletToId`=`WalletObject`.`id`;
E/amplify:flutter:datastore( 6792): 	at android.database.sqlite.SQLiteConnection.nativePrepareStatement(Native Method)
E/amplify:flutter:datastore( 6792): 	at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:986)
E/amplify:flutter:datastore( 6792): 	at android.database.sqlite.SQLiteConnection.prepare(SQLiteConnection.java:593)
E/amplify:flutter:datastore( 6792): 	at android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java:590)
E/amplify:flutter:datastore( 6792): 	at android.database.sqlite.SQLiteProgram.<init>(SQLiteProgram.java:61)
E/amplify:flutter:datastore( 6792): 	at android.database.sqlite.SQLiteQuery.<init>(SQLiteQuery.java:37)
E/amplify:flutter:datastore( 6792): 	at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:46)
E/amplify:flutter:datastore( 6792): 	at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:1443)
E/amplify:flutter:datastore( 6792): 	at android.database.sqlite.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1382)
E/amplify:flutter:datastore( 6792): 	at com.amplifyframework.datastore.storage.sqlite.SQLiteStorageAdapter.getQueryAllCursor(SQLiteStorageAdapter.java:912)
E/amplify:flutter:datastore( 6792): 	at com.amplifyframework.datastore.storage.sqlite.SQLiteStorageAdapter.lambda$query$5$SQLiteStorageAdapter(SQLiteStorageAdapter.java:393)
E/amplify:flutter:datastore( 6792): 	... 6 more
```

From my research earlier trying to solve the problem, it appears that because the TransactionObject model has two different relationships with the WalletObject (from and to), their _WalletObject.id_ fields are being called twice having the same alias. Directly with SQLite, it should be fixed by having different aliases for the field calls but since this is generated by the DataStore plugin, I have no way of modifying them even manually.

I got some idea from [Amplify-Android Issue 908](https://github.com/aws-amplify/amplify-android/issues/908) when I was initially trying to work on the problem but it seems that It does not affect the original aliases.

Here is the sample code I used in relation to the article I linked:

```
await Amplify.DataStore.query(
          TransactionObject.classType,
          where: QueryField(
                      fieldType: WalletObject.ID.fieldType,
                      fieldName: "WalletObjectFromId")
                  .eq('$_walletFromId')
        );
```
Here are some inquiries:

- Is my schema and relationship between Transaction and Wallet (from and to) valid or should I refactor it? or
- Am I missing something instead?

Here's some information about my working environment:

- Flutter 2.0.2
- Dart 2.12.1
- amplify_core: ^0.1.1
- amplify_flutter: ^0.1.1
- amplify_auth_cognito: ^0.1.1
- amplify_datastore: ^0.1.1
- amplify_datastore_plugin_interface: ^0.1.1
- Amplify CLI 4.46.1

**Comment by Amplifiyer** [comment]
I'm guessing this might require some codegen changes on flutter side as well though I'm not sure of the actual root cause. I'm passing this to amplify-android library for further guidance on this.

**Comment by infernal008** [comment]
Hello @Amplifiyer , thanks for checking on this.

I've also read of some updates made with the android-amplify issue I mentioned, but it seems to not work on my case. Maybe something is still amiss with codegen or it I've missed something with my implementation.

**Comment by richardmcclellan** [comment]
Hi infernal008 and @Amplifiyer, there was a similar issue reported here: https://github.com/aws-amplify/amplify-android/issues/908.

The issue is that the `QueryField` object is created using the static `QueryField.field(String fieldName)` method, which doesn't contain the modelName, which can be used to resolve the ambiguity error.

We solved this by adding a new static [`QueryField.field(String modelName, String fieldName)` method](https://github.com/aws-amplify/amplify-android/blob/682c9e778adac74c2261bce4de4866c5e255a5eb/core/src/main/java/com/amplifyframework/core/model/query/predicate/QueryField.java#L49-L57).  The Android Java codegen has been updated to create the models with this method, [here](https://github.com/aws-amplify/amplify-codegen/issues/21).

I believe the solution here is probably just for Flutter to update [this code](https://github.com/aws-amplify/amplify-flutter/blob/c773828dd91ded62e25eece9b8945f07fa551957/packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/query/QueryPredicateBuilder.kt#L37), to use the new method, which allows specification of the modelName.

**Comment by infernal008** [comment]
Hello @richardmcclellan, thanks for sharing your insight to this.

I agree with your remarks and have actually seen the model names be inserted in the generated sql query while running in my local device. I believe it should work fine but my specific use case of calling/relating the same model twice to another model causes ambiguity to the generated queries.

For example:

1. Model1Object has many Model2Objects (Model2Object here treated as a recipient)
2. Model1Object has many Model2Objects (Model2Object here treated as sender)
3. I have used Model2Object twice in different references to Model1Object
4. The joining statement will be something like _SELECT ... FROM Model1Object as Model1Object LEFT JOIN Model2Object as Model2Object on ... LEFT JOIN Model2Object as Model2Object_
5. You can see above that the relationship of Model1Object join with Model2Object twice (due to business logic)
6. The generated sql query did put an alias on the model names but calling the same model again gives it the same alias as it was initially given earlier in the query. Having the same alias for the models in the query causes the ambiguity error.
7. My guess to solve this scenario is if there is a way to give different aliases to model names in the sql query everytime the same model is called in a join statement.
8. Or my modelling may just be wrong or inefficient

Since filing this issue, I have continuously worked on this and have instead opted to change my models in a way that I no longer need the logic requirements as above. It is to say that I have somehow moved on from this problem.

I have just kept this open as  another scenario or business logic may encounter the same issue. But for me, at least, I have been able to go around this problem.

Should this be closed then?

**Comment by richardmcclellan** [comment]
Thanks for the clarification @infernal008!  I see now that my suggested solution probably isn't going to work for your use case, since there is some ambiguity.  We can keep this issue open to track a more long term fix.

**Comment by infernal008** [comment]
Got it, @richardmcclellan! Thanks again and I'll try to help if I could.

**Comment by Jordan-Nelson** [comment]
@richardmcclellan I ran into this issue using `Where.id("my-id")`. It looks like Where.id() is using `QueryField.field(PrimaryKey.fieldName())` and not specifying the model name.

https://github.com/aws-amplify/amplify-android/blob/e3b222d73ebec419f9630869cad127dffc1b88b1/core/src/main/java/com/amplifyframework/core/model/query/Where.java#L59-L63

I was able to work around it by adding a duplicate method that specified the model name like so:
```kt
    private fun whereId(modelName: String, modelId: String): QueryOptions {
        val idField = QueryField.field(modelName, PrimaryKey.fieldName())
        return Where.matches(idField.eq(Objects.requireNonNull(modelId)))
                .paginated(Page.firstResult())
    }
```

Should the current `Where.id()` method be replaced with one that includes the model name?

**Comment by eeatonaws** [comment]
Hi, a fix for the ambiguous column name error when using Where.id has been released as part of Amplify 1.29.0. We are working on a fix for the other ambiguous column name error (querying a model that has 2 columns of the same model type).

**Comment by micstepper** [comment]
@eeatonaws is there a workaround for this at the moment?

**Comment by eeatonaws** [comment]
Hi @micstepper, unfortunately the only workaround at the moment is to modify your schema so no model has two or more fields of the same model type. For example, in the original schema posted in this issue, the TransactionObject could not have 2 WalletObject fields. We are working on a fix for this and will keep you updated here.

**Comment by micstepper** [comment]
@eeatonaws thanks for your response. Any timeline for this fix being released?

**Comment by eeatonaws** [comment]
@micstepper, I apologize for the delay. A fix for the issue has been released as part of Amplify Android 1.31.1.
